ext/ddtrace.c

```diff
@@ -1185,6 +1191,71 @@ static PHP_FUNCTION(add_distributed_tag) {
     RETURN_NULL();
 }
 
+static PHP_FUNCTION(set_user) {
+    UNUSED(execute_data);
+
+    zend_string *user_id = NULL;
+    HashTable *metadata = NULL;
+    zend_bool propagate = false;
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "S|hb", &user_id, &metadata,
+      &propagate) == FAILURE) {
+        ddtrace_log_debug("Unexpected parameter combination, expected "
+                          "(user_id, metadata, propagate)");
+        RETURN_NULL();
+    }
+
+    if (!get_DD_TRACE_ENABLED()) {
+        RETURN_NULL();
+    }
+
+    if (user_id == NULL || ZSTR_LEN(user_id) == 0) {
+        ddtrace_log_debug("Unexpected empty user id");
+        RETURN_NULL();
+    }
+
+    zend_array *target_table;
+    if (DDTRACE_G(active_stack)->root_span) {
+        target_table = ddtrace_spandata_property_meta(DDTRACE_G(active_stack)->root_span);
+    } else {
+        target_table = &DDTRACE_G(root_span_tags_preset);
+    }
+
+    zval user_id_zv;
+    ZVAL_STR_COPY(&user_id_zv, user_id);
+    zend_hash_str_update(target_table, ZEND_STRL("usr.id"), &user_id_zv);
+
+    if (propagate) {
+        zval value_zv;
+        ZVAL_COPY(&value_zv, &user_id_zv);
+        zend_hash_str_update(target_table, ZEND_STRL("_dd.p.usr.id"), &value_zv);
+
+        zend_hash_str_add_empty_element(&DDTRACE_G(propagated_root_span_tags),
+            ZEND_STRL("_dd.p.usr.id"));
+    }
+
+    if (metadata != NULL) {
+        zend_string *key = NULL;
+        zval *value = NULL;
```

Comment:
I'm a proponent of trying to assign default values as little as possible so that a compiler may warn about possibly uninitialized values instead of running into a NULL at runtime.

Reply:
Fair enough I'll follow that convention where feasible.